### ISSUE TITLE
Best practice for JS Singleton pattern

### ISSUE THREAD

**Issue by thisismydesign** [c-hive/guides#39]
It isn't obvious how to do it right: https://stackoverflow.com/questions/1479319/simplest-cleanest-way-to-implement-singleton-in-javascript

We're using the [ES6 class approach](https://stackoverflow.com/a/46184368/2771889) but it would be better if we'd use `.instance` or `.instance()` to make the pattern obvious for readers. Also, when to freeze the object should also be clarified.

**Comment by gomorizsolt** [comment]
Yea, seems to be a controversial topic. How about this approach btw?

```js
class Singleton {
   static instance = null;

   static getInstance() {
      if (!Singleton.instance) {
         // Foo is the class we'd like to be singleton.
         Singleton.instance = new Foo();
      }

      return Singleton.instance;
   }
}

const s1 = Singleton.getInstance();
const s2 = Singleton.getInstance();

console.log(s1 === s2);
// => true, because both classes point to the same reference
```

It's akin to [java's approach](https://www.geeksforgeeks.org/singleton-class-java/) despite the fact `Singleton` can be instantiated because there's no [obvious](https://stackoverflow.com/questions/21667149/how-to-define-private-constructors-in-javascript) way to make constructors private in js.

**Comment by gomorizsolt** [comment]
What's the matter with the above-mentioned proposal? Would you like to apply it to the projects where it's relevant? Or let's hold off this decision?

**Comment by thisismydesign** [comment]
Note, there are also private static fields: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_fields

**Comment by gomorizsolt** [comment]
Yes, but that's still an experimental feature(stage 3), see https://github.com/tc39/proposal-class-fields.

**Comment by thisismydesign** [comment]
Both public and private though:

> Both Public and private field declarations are an experimental feature (stage 3)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_fields

**Comment by thisismydesign** [comment]
So we switched to another approach that seems to be working well. @gomorizsolt could you add it to the guides and resolve this issue?

**Comment by gomorizsolt** [comment]
Whoa, I can't recall this topic anymore. What's our preferred approach right now?

```js
class Singleton {
   static instance = null;

   static getInstance() {
      if(!this.instance) {
         this.instance = new Foo();
      }

      return this.instance;
   }
}
```

AFAICT it's identical to https://github.com/c-hive/guides/issues/39#issuecomment-575040514.

**Comment by thisismydesign** [comment]
Whatever we use in Sniper should be the currently preferred approach.

**Comment by gomorizsolt** [comment]
Resolved by #43.
